### ISSUE TITLE
"Post-installation script subprocess returned error" on Ubuntu 18.04

### ISSUE THREAD

**Issue by joskapista44** [ONLYOFFICE/DocumentServer#733]
**Do you want to request a *feature* or report a *bug*?**
bug
**What is the current behavior?**
software update fails with an error
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
I have an Ubuntu 18.04 server, with an olnyoffice 5.4.1-39. When I try to upgrade it to the next version (5.4.2-46), via APT, I get an error:

    dpkg: error processing package onlyoffice-documentserver (--configure):
    installed onlyoffice-documentserver package post-installation script subprocess returned error exit status 1
    Errors were encountered while processing:
    onlyoffice-documentserver
    E: Sub-process /usr/bin/dpkg returned an error code (1)

When I try to configure it:

    sudo dpkg --configure -a
    [sudo] password for user:
    Setting up onlyoffice-documentserver (5.4.2-46) ...
    Generating AllFonts.js, please wait...Done
    Generating presentation themes, please wait...Done
    dpkg: error processing package onlyoffice-documentserver (--configure):
    installed onlyoffice-documentserver package post-installation script subprocess returned error exit status 1
    Errors were encountered while processing:
    onlyoffice-documentserver

**What is the expected behavior?**
updating without any problem

**Did this work in previous versions of DocumentServer?**
I had the same problem earlier

**DocumentServer version:**
olnyoffice 5.4.1-39. upgrade to 5.4.2-46

**Operating System:**
Ubuntu 18.04 server (up to date)

**Comment by ShockwaveNN** [comment]
@joskapista44 Hi, please make sure that Postgresql, rabbitmq and redis are started and working correctly (no errors in logs)

**Comment by joskapista44** [comment]
Hey ShockwaveNN, thanks for tip.
So before I try the upgrade all of them looks fine. I could find only a minor log entry in the Postgresql log:
"2020-01-07 10:42:55.268 UTC 11551 [unknown] [unknown] LOG:  incomplete startup packet"
I tried to find any solution to resolve it, but read everywhere it's harmless: "this message is printed in the log if a client makes a TCP connection to the PG server, but sends no traffic". I'm not a professional but I don't think this is the source of the problem.

**Comment by ShockwaveNN** [comment]
@joskapista44 Try to run `bash /var/lib/dpkg/info/onlyoffice-documentserver.postinst`

**Comment by joskapista44** [comment]
I run it and got this:
Use of uninitialized value $action in string eq at /usr/share/debconf/frontend line 27.
postinst called with unknown argument $1

These are the first few rows of the frontend, I did not change anything:

>  #!/usr/bin/perl -w
> # This file was preprocessed, do not edit!
> 
> use strict;
> use Debconf::Db;
> use Debconf::Template;
> use Debconf::AutoSelect qw(:all);
> use Debconf::Log qw(:all);
> 
> Debconf::Db->load;
> 
> debug developer => "frontend started";
> 
> my $frontend=make_frontend();
> 
> shift @ARGV if $ARGV[0] eq '--';
> 
> my $package;
> my $no_title=0;
> if ($ENV{DEBCONF_PACKAGE}) {
>         $package=$ENV{DEBCONF_PACKAGE};
> }
> elsif ($ARGV[0]=~m!^.*/(.*?)\.(?:postinst|postrm|prerm)$!) {
>         $package=$1;
>         my $action=$ARGV[1];
>         $no_title=1 if $action eq 'triggered';
> }
>

**Comment by ShockwaveNN** [comment]
@joskapista44 Sorry, my mistake, try use those command to get detailed error logs
https://github.com/ONLYOFFICE/DocumentServer/issues/478#issuecomment-455586822

**Comment by joskapista44** [comment]
Still in trouble..
1. npm is already installed
2. changed the first line of /var/lib/dpkg/info/onlyoffice-documentserver.postinst to # !/bin/sh -x
and run /var/lib/dpkg/info/onlyoffice-documentserver.postinst configure from root user
[result1.txt](https://github.com/ONLYOFFICE/DocumentServer/files/4043630/result1.txt)
3.  tried to clean everything up (/etc/onlyoffice, /var/www/onlyoffice) and reinstall but can't uninstall onlyoffice, I get the same error like I try to update it.

**Comment by ShockwaveNN** [comment]
@joskapista44 This is strange, that `/var/lib/dpkg/info/onlyoffice-documentserver.postinst configure` run without error.
If you got troubles removing - you can remove `/var/lib/dpkg/info/onlyoffice-documentserver.prerm` and `/var/lib/dpkg/info/onlyoffice-documentserver.postrm` files and uninstall again.

But in general - we always recommend to use Docker version

**Comment by joskapista44** [comment]
Yes different result with the two methods. ("/var/lib/dpkg/info/onlyoffice-documentserver.postinst configure" VS "bash /var/lib/dpkg/info/onlyoffice-documentserver.postinst")
With the version 4 I can remove it, but after I can't install, so still the same result

**Comment by ShockwaveNN** [comment]
> but after I can't install

With which error? I think there may be some troubles with connection to postgresql, check that connection settings are correct

**Comment by joskapista44** [comment]
I wanted to send you the error message, so I purged the onlyoffice-documentserver, and deleted the folders as a few times I did before. I don't know what was now different but after the removing now I was able to reinstall it without problems, and now everything works. The only thing is that this is not a solution of the original problem. I'm sure I will have the same problem with the next upgrade. So :) and a bit :(

**Comment by ShockwaveNN** [comment]
I agree, that problem still not solved. Previously we got several reports
with se problem, but no one able to provide setup in which I can reproduce
the issue. So I do not what to do either, but I think I close issue until
someone provide exact steps to reproduce

**Comment by joskapista44** [comment]
Yes I can accept, it's not easy to catch the problem. But my situation looks reproducible.
I have an 18.04 ubuntu server on a virtualbox vm. I followed this tutorial to setup it (I don't know whether its important or not): https://www.digitalocean.com/community/tutorials/how-to-install-and-configure-owncloud-on-ubuntu-18-04
Then I install the onlyoffice with ngingx. (https://helpcenter.onlyoffice.com/server/linux/document/linux-installation.aspx)
So everything works well, till the first upgrade of onlyoffice. I had the same issue with ubuntu 16.04 server and a previous version of OO. I'm sure if I install the same setup again I would have the same issue.

**Comment by ShockwaveNN** [comment]
@joskapista44 Did you setup owncloud and Documentserver on same server? seems this will result that you'll have apache2 and nginx on same server and this is not a good thing

**Comment by joskapista44** [comment]
Yes, and they operates without any issues on different ports. I'm not a pro but I couldn't see any problem with it

**Comment by ShockwaveNN** [comment]
@joskapista44 Could you provide vm image for us to reproduce the problem?

**Comment by joskapista44** [comment]
Sure. So I started to rebuild the clone of my whole server form the beginning. I started with onlyoffice document server 5.4.0-21, from a deb package. After I finished the setup of the onlyoffice I made a snapshot backup and tried to upgrade it to 5.4.2-46 from repository. It worked. 
Went back to the snapshot and I started to install a LAMP and owncloud. The first step was the apache2. After apache2 install, I tried to upgrade again and I got the known bug. So the problem must be there. I can share you both snapshots if you still need. Can I contact you directly?

**Comment by ShockwaveNN** [comment]
@joskapista44 yeah, you can contact me via shockwavenn@gmail.com

**Comment by ShockwaveNN** [comment]
@joskapista44 send me a VM image with bug reproduceses

I changed `/var/lib/dpkg/info/onlyoffice-documentserver.postinst` - changed to `#!/bin/sh -x`
according to https://askubuntu.com/a/735568/39452 

and run 
`sudo dpkg --configure -a`

and script stopped at 
`service nginx restart`

There was errors in `systemctl status nginx.service`
```
nginx: [emerg] bind() to 0.0.0.0:80 failed (98: Address already in use)
```

I stopped apache2 
`sudo service apache2 stop`
and run upgrade process again - everything is fine

Now I'll try to figure out why nginx is broken during update

**Comment by ShockwaveNN** [comment]
seems that port of DocumentServer was changed manually by editing nginx config, not by [instruction](https://helpcenter.onlyoffice.com/server/linux/document/linux-installation.aspx)

This value is unchanged and used in upgrade process
```
sudo debconf-show onlyoffice-documentserver | grep ds-port
* onlyoffice/ds-port: 80
```

**Comment by joskapista44** [comment]
So if I apply this with my actual server it will solve the problem?

> echo onlyoffice-documentserver onlyoffice/ds-port select <PORT_NUMBER> | sudo debconf-set-selections

**Comment by ShockwaveNN** [comment]
@joskapista44 You've got pretty complicated setup, I'm not sure everything will be fine, but I'm sure without this after each update DocumentServer will be reverted to 80 port

**Comment by kwisatz** [comment]
I know this ticket is closed, just wanted to chime in, say thanks for the hints on how to investigate this and that I had a similar problem (also nextcloud and onlyoffice documentserver on the same machine).
postinst failed at: ` service nginx restart`

The reason in my case seems to have been that `nginx` had already been removed from the system (so I'm guessing that maybe there is a missing dependency on it and that uninstalling `nginx` should have also prompted/forced uninstallation of the documentserver ?)

**Comment by ShockwaveNN** [comment]
> The reason in my case seems to have been that nginx had already been removed from the system

I think this may be a case but I would be thankful if you provide some cases steps to reproduce this, except that ports of DocumentServer was setup incorrectly
I think we can have some bug in our init scripts as well but I was never able to reproduce it more than once

And generally speaking - setup both DocumentServer and Nextcloud on same server not a great idea, it's better to use dockers - I think it's simpler and more robust

**Comment by Alg0v** [comment]
This issue is a bit old, but I had the same issue on my server. I think that the problem is that the script tries to run OnlyOffice on the port 80, but on some machines like mine, there's already something running on the port 80, so it crashes. I fixed that issue by stopping Apache in my case (with `systemctl stop apache2`).
